pkg/dockerregistry/server/signaturedispatcher.go

```diff
@@ -0,0 +1,187 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+
+	ctxu "github.com/docker/distribution/context"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	"github.com/docker/distribution/registry/handlers"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+
+	gorillahandlers "github.com/gorilla/handlers"
+)
+
+const (
+	errGroup             = "registry.api.v2"
+	defaultSchemaVersion = 2
+)
+
+// signature represents a Docker image signature.
+type signature struct {
+	Version int `json:"schemaVersion"`
+	// Name must be in "sha256:<digest>@signatureName" format
+	Name string `json:"name"`
+	// Type is optional, of not set it will be defaulted to "AtomicImageV1"
+	Type string `json:"type"`
+	// Content contains the signature
+	Content []byte `json:"content"`
+}
+
+// signatureList represents list of Docker image signatures.
+type signatureList struct {
+	Signatures []signature `json:"signatures"`
+}
+
+var (
+	ErrorCodeSignatureInvalid = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_INVALID",
+		Message:        "invalid image signature",
+		HTTPStatusCode: http.StatusBadRequest,
+	})
+
+	ErrorCodeSignatureAlreadyExists = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_EXISTS",
+		Message:        "image signature already exists",
+		HTTPStatusCode: http.StatusConflict,
+	})
+)
+
+type signatureHandler struct {
+	ctx       *handlers.Context
+	reference imageapi.DockerImageReference
+}
+
+// SignatureDispatcher handles the GET and PUT requests for signature endpoint.
+func SignatureDispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
+	signatureHandler := &signatureHandler{ctx: ctx}
+	signatureHandler.reference, _ = imageapi.ParseDockerImageReference(ctxu.GetStringValue(ctx, "vars.name") + "@" + ctxu.GetStringValue(ctx, "vars.digest"))
+
+	return gorillahandlers.MethodHandler{
+		"GET": http.HandlerFunc(signatureHandler.Get),
+		"PUT": http.HandlerFunc(signatureHandler.Put),
+	}
+}
+
+func (s *signatureHandler) Put(w http.ResponseWriter, r *http.Request) {
+	context.GetLogger(s.ctx).Debugf("(*signatureHandler).Put")
+	if len(s.reference.String()) == 0 {
+		s.handleError(s.ctx, v2.ErrorCodeNameInvalid.WithDetail("the image ID must be specified (sha256:<digest>"), w)
```

Comment:
This message looks like this in the console:
```
{"errors":[{"code":"NAME_INVALID","message":"invalid repository name","detail":"the image ID must be specified (sha256:\u003cdigest\u003e"}]}
```
Maybe we should use ASCII only ?